LibreHardwareMonitorLib/Hardware/Memory/MemoryGroup.cs

```diff
@@ -15,132 +18,174 @@
     {
-        if (Ring0.IsOpen)
+        if (Ring0.IsOpen && (DriverManager.Driver is null || !DriverManager.Driver.IsOpen))
         {
             // Assign implementation of IDriver.
             DriverManager.Driver = new RAMSPDToolkitDriver(Ring0.KernelDriver);
             SMBusManager.UseWMI = false;
         }
-        else
+
+        _hardware.Add(new VirtualMemory(settings));
+        _hardware.Add(new TotalMemory(settings));
+
+        if (DriverManager.Driver == null)
         {
-            // Still need to set Driver if Ring0 is absent.
-            DriverManager.Driver = new RAMSPDToolkitDriver(null);
-            SMBusManager.UseWMI = false;
+            return;
+        }
+
+        if (!TryAddDimms(settings))
+        {
+            StartRetryTask(settings);
         }
+
+        _opened = true;
     }
 
-    public MemoryGroup(ISettings settings)
+    private void StartRetryTask(ISettings settings)
     {
-        _hardware.Add(new VirtualMemory(settings));
-        _hardware.Add(new TotalMemory(settings));
+        _cancellationTokenSource = new CancellationTokenSource();
 
-        //No RAM detected
-        if (!DetectThermalSensors(out List<SPDAccessor> accessors))
+        Task.Run(async () =>
         {
-            //Retry a couple of times
-            //SMBus might not be detected right after boot
-            _timer = new Timer(RetryTime);
+            int retryRemaining = 5;
 
-            _timer.Elapsed += (_, _) =>
+            while (!_cancellationTokenSource.IsCancellationRequested && --retryRemaining > 0)
             {
-                if (_elapsedCounter++ >= RetryCount || DetectThermalSensors(out accessors))
+                await Task.Delay(TimeSpan.FromSeconds(2.5), _cancellationTokenSource.Token);
+
+                if (TryAddDimms(settings))
                 {
-                    _timer.Stop();
-                    _timer = null;
+                    lock (_lock)
+                    {
+                        if (!_opened)
+                        {
+                            return;
+                        }
 
-                    if (accessors != null)
-                        AddDimms(accessors, settings);
-                }
-            };
+                        foreach (Hardware hardware in _hardware)
+                        {
+                            HardwareAdded?.Invoke(hardware);
+                        }
 
-            _timer.Start();
-        }
-        else
-        {
-            AddDimms(accessors, settings);
-        }
+                        _cancellationTokenSource.Dispose();
+                        _cancellationTokenSource = null;
+
+                        break;
+                    }
+
+                }
+            }
+        }, _cancellationTokenSource.Token);
     }
 
     public IReadOnlyList<IHardware> Hardware => _hardware;
 
     public string GetReport()
     {
-        return null;
+        return _lastException?.ToString() ?? null;
```

Comment:
Remove or make this an actual report, this is used for computer reports.

Reply:
Will make it a full report , might as well.